### PR TITLE
sbcl: remove patches, enable checks, allow overriding

### PR DESCRIPTION
**Pull request by hraban**
## Description of changes

This is a mix of various improvements and refactoring to the SBCL build. Every change is in its own commit, I can create separate small PRs if that's preferred.

Notably:
- Disable the timestamp patches on SBCL which seem to be unnecessary; there are no nix store violation from rebuilding zero-timestamped files, so these patches don't seem to be solving anything that was broken.
- enable `doCheck = true`.
- liberally disable any individual test that causes a problem, without worrying too much about why. It's an improvement over what we had before (which was no tests run at all).

(Ignore the branch name)

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by 7c6f434c** [comment]
Yes, it is good to have!

File: pkgs/development/compilers/sbcl/2.x.nix [hunk 1]
```diff
@@ -41,29 +41,13 @@ stdenv.mkDerivation rec {
...
-    sed -e '5,$d' -i contrib/sb-bsd-sockets/tests.lisp
-    sed -e '5,$d' -i contrib/sb-simple-streams/*test*.lisp
-  ''
-  + (if purgeNixReferences
+  postPatch = (lib.optionalString (builtins.elem stdenv.hostPlatform.system [
+    "x86_64-linux"
+    "x86_64-darwin"
+    "aarch64-linux"
+  ]) ''
+    rm tests/foreign-stack-alignment.impure.lisp
```
Comment (hraban) [inline pkgs/development/compilers/sbcl/2.x.nix:49]: This test fails on x86_64-darwin, but only when run from within a `nix develop ...#sbcl`. Not if I just build sbcl directly on the host. So there's something nixpkgs-specific going on.. My OS is clang 14; I've tried from nixpkgs with clang 11 (default) and 16 and neither worked ( I guess I could try with 14 but I assume at this point that won't make a difference). I'm not sure what else is different.. maybe which zlib is being linked for core compression?

To complicate matters further: this test works fine on aarch64-darwin. The plot thickens!

**Comment by hraban** [comment]
@ofborg build sbcl

**Comment by hraban** [comment]
@ofborg eval

**Comment by hraban** [comment]
@ofborg build sbcl

**Comment by hraban** [comment]
For some strange reason the failing tests on borg seem to be rebuilding a ton of unrelated packages at which point it runs out of space. The only one that didn't attempt to do that was the original aarch64-darwin run, which subsequently passed.

**Comment by 7c6f434c** [comment]
Not really unrelated, just pretty deep dependencies… Staging is a bit of a cointoss with that.

**Comment by hraban** [comment]
> Not really unrelated, just pretty deep dependencies… Staging is a bit of a cointoss with that.

Ah, does that mean ofborg builds the merge target, not the branch head?

**Comment by 7c6f434c** [comment]
Yes, it does a local merge.

**Comment by 7c6f434c** [comment]
@ofborg build sbcl.pkgs.stumpwm

**Comment by 7c6f434c** [comment]
@ofborg build sbcl.pkgs.stumpwm

**Comment by hraban** [comment]
Weird, it can't find `env` in `PATH` even though I've added `coreutils` to `nativeBuildInputs`, and that same solution has worked for all other programs needed during check (`strace`, `which`, ...)

**Comment by 7c6f434c** [comment]
I wonder if `which` would find `env`… A pity that a lot of stuff needs to be loaded in order so using `strace` to figure out the search process has horrible overhead.

**Comment by hraban** [comment]
I'd love an easy way to run those ofborg tests locally in e.g. a docker container

**Comment by hraban** [comment]
Well, `env` is clearly in the `PATH`: "https://logs.ofborg.org/?key=nixos/nixpkgs.262107&attempt_id=4ba24723-0121-4e76-b1dc-e087c3a8d1f9". And why wouldn't it be...

I'll keep debugging some more but it's a very frustrating process because I have a hard time reproducing an ofborg runner. It'd be nice if there were an .iso or a nixos config or something like that which I could just load locally to play around with :/

**Comment by 7c6f434c** [comment]
While fully reproducing ofborg checks is complicated, here it is just `nix-build './nixpkgs-pr-checkout' -A sbcl` that fails. Can reproduce locally on x86_64-linux, would expect the same to happen with a stock NixOS VM. (I am using Nixpkgs kernel and Nixpkgs Nix with not that exotic of a configuration, so sandboxing should work the same).

A build does take a long time, indeed…

(I wonder if ECL would be faster or slower)

**Comment by hraban** [comment]
> While fully reproducing ofborg checks is complicated, here it is just `nix-build './nixpkgs-pr-checkout' -A sbcl` that fails.

Interesting. Where is that `nixpkgs-pr-checkout` file?

> Can reproduce locally on x86_64-linux, would expect the same to happen with a stock NixOS VM. (I am using Nixpkgs kernel and Nixpkgs Nix with not that exotic of a configuration, so sandboxing should work the same).

Fascinating--I cannot repro. It finds `env` just fine here on a relatively simple nixos VM. I do get a build error, but it's an unrelated failure in solist.pure.lisp (which doesn't happen on CI 🙄)

> A build does take a long time, indeed…
> 
> (I wonder if ECL would be faster or slower)

Tangent: with `bootstrapLisp` being overridable you can do:

```
nix build --print-out-paths --no-link --impure --expr 'let realpkgs = import <nixpkgs> {}; devpkgs = import  ~/my/dev/nixpkgs {}; in devpkgs.sbcl.override { bootstrapLisp = "${realpkgs.sbcl}/bin/sbcl --disable-debugger --no-userinit --no-sysinit"; } '
```

Which is much faster.

**Comment by 7c6f434c** [comment]
> Interesting. Where is that `nixpkgs-pr-checkout` file?

No-no this is just one of the multiple Nixpkgs checkouts that I have side by side.

> Fascinating--I cannot repro using bare `nix-build -A sbcl`. It finds `env` just fine here on a relatively simple nixos VM.

This is via Rosetta? Although it should not be easy to affect this…

> Tangent: with `bootstrapLisp` being overridable you can do:

Well, it is also good to have faster Hydra builds afterwards!  (But thanks)

**Comment by hraban** [comment]
> > Fascinating--I cannot repro using bare `nix-build -A sbcl`. It finds `env` just fine here on a relatively simple nixos VM.
> 
> This is via Rosetta? Although it should not be easy to affect this…

No I'm running x86_64 NixOS in an emulated qemu VM via UTM (host is aarch64-darwin yes, but no rosetta, this is emulation)

Is it actually the same error you're getting, that shell script (the test ending in `.sh`)? Or is a `.lisp` test that is failing for you?

> > Tangent: with `bootstrapLisp` being overridable you can do:
> 
> Well, it is also good to have faster Hydra builds afterwards! (But thanks)

Fair. I tried ECL back when I introduced source only building but it wasn't as cross-platform reliable as CLISP so I gave up. Could have another look.

Do you happen to know how golang does it? Iirc their compiler is bootstrapped using version `n-1` (iow the source code for the compiler is allowed to use features that were introduced one minor version ago), so you get this long chain of compilers from a version long-long ago that was written in C. If they have a nice model for that we could technically do the same with sbcl...?

**Comment by 7c6f434c** [comment]
Oops sorry, accidentally restarted the build (which loses logs…)

So, I build in commit  2c4c3e4cc14dd37f00cb11df111cad46b05236ca and this means derivation /nix/store/qlhj70qazk5dvrihlxkq9kk2n631nl5y-sbcl-2.4.1.drv will report (in the CET-late-morning, I guess)

Go uses either `gccgo` or upstream binaries. (Just looked in their expressions)

**Comment by 7c6f434c** [comment]
```
Failure:            type.pure.lisp / (TYPE-DERIVATION LOGICAL-OPERATIONS SCALING)
Invalid exit status: run-program.test.sh
```

**Comment by 7c6f434c** [comment]
Just in case: is your test passing in VM the same exact derivation path (i.e. the same hash part)?

**Comment by hraban** [comment]
```
sbcl> Status:
sbcl>  Skipped (broken):   gethash-concurrency.pure.lisp / (HASH-TABLE UNSYNCHRONIZED)
sbcl>  Skipped (broken):   hash.pure.lisp / (HASH-TABLE GC-SMASHED-CELL-LIST)
sbcl>  Expected failure:   hash.pure.lisp / SXHASH-ON-DISPLACED-STRING
sbcl>  Failure:            solist.pure.lisp / CONCURRENTLY-INSERT-SAME-KEYS/STRING
sbcl>  Failure:            solist.pure.lisp / CONCURRENTLY-INSERT-SAME-KEYS/OBJECT
sbcl>  Expected failure:   block-compile.impure.lisp / BLOCK-COMPILE-TOP-LEVEL-CLOSURES.SAME-ENVIRONMENT.LOCAL-CALLS
sbcl>  Expected failure:   compiler-2.impure.lisp / TOP-LEVEL-CLOSURE-SEPARATE-COMPONENT
sbcl>  Expected failure:   compiler-2.impure.lisp / TOP-LEVEL-CLOSURE-SEPARATE-COMPONENT.2
sbcl>  Expected failure:   full-eval.impure.lisp / INLINE-FUN-CAPTURES-DECL
sbcl>  Expected failure:   gc.impure.lisp / M-A-O-THREADLOCALLY-PRECISE
sbcl>  Expected failure:   gc.impure.lisp / REPEATABLY-COUNT-ALLOCATED-OBJECTS
sbcl>  Expected failure:   gc.impure.lisp / PIN-ALL-CODE-WITH-GC-ENABLED
sbcl>  Skipped (broken):   gc.impure.lisp / CODE-ITERATION-FAST
sbcl>  Skipped (broken):   gc.impure.lisp / PAGE-PROTECTED-P
sbcl>  Expected failure:   packages.impure.lisp / USE-PACKAGE-CONFLICT-SET
sbcl>  Expected failure:   packages.impure.lisp / IMPORT-SINGLE-CONFLICT
sbcl>  Skipped (broken):   run-program.impure.lisp / (RUN-PROGRAM AUTOCLOSE-STREAMS)
sbcl>  Expected failure:   traceroot.impure.lisp / (SEARCH-ROOTS STACK-INDIRECT)
sbcl>  Expected failure:   traceroot.impure.lisp / TRACEROOT-COLLAPSE-LISTS
sbcl>  (38 tests skipped for this combination of platform and features)
sbcl> test failed, expected 104 return code, got 1
error: builder for '/nix/store/qlhj70qazk5dvrihlxkq9kk2n631nl5y-sbcl-2.4.1.drv' failed with exit code 1;
```

Weird!

**Comment by 7c6f434c** [comment]
Hmmmmmm. It was a NixOS VM with sandboxing on and no funny stuff configured like `/usr/bin/env` in the sandbox?

**Comment by 7c6f434c** [comment]
Good news: I know why the test fails. Bad news: I have no idea why it works on your machine.

Basically when `:environment` is passed it is set then `execvp` is called. Yep, `PATH` is taken from the override environment which only contains a single intentionally-nonsense variable. `execvp` still tried `/bin` `/usr/bin` `/run/current-system/sw/bin` but none of these contain `env`.

**Comment by 7c6f434c** [comment]
(replacing `env` with an absolute path to `env` seems to work)

**Comment by hraban** [comment]
> Good news: I know why the test fails. Bad news: I have no idea why it works on your machine.
> 
> Basically when `:environment` is passed it is set then `execvp` is called. Yep, `PATH` is taken from the override environment which only contains a single intentionally-nonsense variable. `execvp` still tried `/bin` `/usr/bin` `/run/current-system/sw/bin` but none of these contain `env`.

OHHHHHHHH MY DAYS of course oh man that is so obvious thank you. Ok I will fix this.

Meanwhile, my machine strays ever further from God's path:

```
sbcl> ::: Running :ADDRESS-BASED-SXHASH-GCING
sbcl> CORRUPTION WARNING in SBCL pid 957 tid 957:
sbcl> Memory fault at 0x3f0 (pc=0x44ab97, fp=0x7ffff417e5b0, sp=0x7ffff417e4e0) tid 957
sbcl> The integrity of this image is possibly compromised.
sbcl> Exiting.
sbcl>    0: fp=0x7ffff417e5b0 pc=0x44ab97 Foreign function run_compaction
sbcl>    1: fp=0x7ffff417e610 pc=0x449857 Foreign function mr_collect_garbage
sbcl>    2: fp=0x7ffff417e670 pc=0x43a4b9 Foreign function garbage_collect_generation
sbcl>    3: fp=0x7ffff417e720 pc=0x43ae1b Foreign function collect_garbage
sbcl>    4: fp=0x7ffff417e738 pc=0x10011c83a7 SB-KERNEL::COLLECT-GARBAGE
sbcl>    5: fp=0x7ffff417e7d8 pc=0x10011c8bc4 (FLET "WITHOUT-GCING-BODY-5" :IN SB-KERNEL::SUB-GC)
sbcl>    6: fp=0x7ffff417e878 pc=0x10011c891d (FLET "WITHOUT-INTERRUPTS-BODY-1" :IN SB-KERNEL::SUB-GC)
sbcl>    7: fp=0x7ffff417e918 pc=0x10011c86fc SB-KERNEL::SUB-GC
sbcl>    8: fp=0x7ffff417e938 pc=0x10011c913e SB-EXT::GC
sbcl>    9: fp=0x7ffff417e9a0 pc=0x1001f9e50d (LAMBDA () :IN "/build/sbcl-2.4.1/tests/hash-2.pure.lisp")
sbcl>   10: fp=0x7ffff417ebb0 pc=0x1001f5a9e5 TEST-UTIL::RUN-TEST
sbcl>   11: fp=0x7ffff417ec88 pc=0x1001069415 SB-INT::SIMPLE-EVAL-IN-LEXENV
sbcl>   12: fp=0x7ffff417ecb0 pc=0x1001069b39 SB-EXT::EVAL-TLF
sbcl>   13: fp=0x7ffff417ef08 pc=0x100126a012 (LABELS SB-FASL::EVAL-FORM :IN SB-INT::LOAD-AS-SOURCE)
sbcl>   14: fp=0x7ffff417f020 pc=0x1001269935 (LAMBDA (SB-KERNEL::FORM &KEY :CURRENT-INDEX &ALLOW-OTHER-KEYS) :IN SB-INT::LOAD-AS-SOURCE)
sbcl>   15: fp=0x7ffff417f0d0 pc=0x10007c8812 SB-C::%DO-FORMS-FROM-INFO
sbcl>   16: fp=0x7ffff417f2e0 pc=0x100126920e SB-INT::LOAD-AS-SOURCE
sbcl>   17: fp=0x7ffff417f3f8 pc=0x100126bcc7 (LABELS SB-FASL::LOAD-STREAM-1 :IN LOAD)
sbcl>   18: fp=0x7ffff417f490 pc=0x100126a6ac SB-FASL::CALL-WITH-LOAD-BINDINGS
sbcl>   19: fp=0x7ffff417f5a8 pc=0x100126b74c LOAD
sbcl>   20: fp=0x7ffff417f600 pc=0x10029aa23b RUN-TESTS::LOAD-TEST
sbcl>   21: fp=0x7ffff417f7f8 pc=0x1004909ea9 RUN-TESTS::PURE-RUNNER
sbcl>   22: fp=0x7ffff417f890 pc=0x10029abbcd RUN-TESTS::RUN-ALL
sbcl>   23: fp=0x7ffff417f968 pc=0x1001069415 SB-INT::SIMPLE-EVAL-IN-LEXENV
sbcl>   24: fp=0x7ffff417f980 pc=0x100106993d EVAL
sbcl>   25: fp=0x7ffff417fb90 pc=0x100106c5fe SB-IMPL::PROCESS-EVAL/LOAD-OPTIONS
sbcl>   26: fp=0x7ffff417fdd8 pc=0x100106e77c SB-IMPL::TOPLEVEL-INIT
sbcl>   27: fp=0x7ffff417fe70 pc=0x1001496f91 (FLET SB-UNIX::BODY :IN SB-IMPL::START-LISP)
sbcl>   28: fp=0x7ffff417ff30 pc=0x1001496d8f (FLET "WITHOUT-INTERRUPTS-BODY-3" :IN SB-IMPL::START-LISP)
sbcl>   29: fp=0x7ffff417ffc8 pc=0x1001496bb8 SB-IMPL::%START-LISP
sbcl> test failed, expected 104 return code, got 1
error: builder for '/nix/store/qlhj70qazk5dvrihlxkq9kk2n631nl5y-sbcl-2.4.1.drv' failed with exit code 1;
       last 10 log lines:
       >   21: fp=0x7ffff417f7f8 pc=0x1004909ea9 RUN-TESTS::PURE-RUNNER
       >   22: fp=0x7ffff417f890 pc=0x10029abbcd RUN-TESTS::RUN-ALL
       >   23: fp=0x7ffff417f968 pc=0x1001069415 SB-INT::SIMPLE-EVAL-IN-LEXENV
       >   24: fp=0x7ffff417f980 pc=0x100106993d EVAL
       >   25: fp=0x7ffff417fb90 pc=0x100106c5fe SB-IMPL::PROCESS-EVAL/LOAD-OPTIONS
       >   26: fp=0x7ffff417fdd8 pc=0x100106e77c SB-IMPL::TOPLEVEL-INIT
       >   27: fp=0x7ffff417fe70 pc=0x1001496f91 (FLET SB-UNIX::BODY :IN SB-IMPL::START-LISP)
       >   28: fp=0x7ffff417ff30 pc=0x1001496d8f (FLET "WITHOUT-INTERRUPTS-BODY-3" :IN SB-IMPL::START-LISP)
       >   29: fp=0x7ffff417ffc8 pc=0x1001496bb8 SB-IMPL::%START-LISP
       > test failed, expected 104 return code, got 1
       For full logs, run 'nix log /nix/store/qlhj70qazk5dvrihlxkq9kk2n631nl5y-sbcl-2.4.1.drv'.
```

I will try again I guess.

**Comment by 7c6f434c** [comment]
> OHHHHHHHH MY DAYS of course oh man that is so obvious thank you. Ok I will fix this.

I must admit I did miss it too when looking at the test, so I had to run strace (and I have messed up the patching on the first try…) and then, knowing what exactly is going on, compare with other tests and then still track in the source code how `:search t` is implemented. There are other ways to implement it where the problem wouldn't appear!

> Meanwhile, my machine strays ever further from God's path:

Why am I wondering whether Apple interprets the Aarch64 ISA in a way Qemu's x86_64 emulation code does not expect?…

**Comment by hraban** [comment]
image
image

2 down, 2 to go.

File: pkgs/development/compilers/sbcl/default.nix [hunk 2]
```diff
@@ -78,48 +85,63 @@ stdenv.mkDerivation rec {
     inherit (versionMap.${version}) sha256;
   };
 
-  nativeBuildInputs = [ texinfo ];
+  nativeBuildInputs = [
+    texinfo
+  ] ++ lib.optionals self.doCheck (
+    [
+      which
+    ] ++ lib.optionals (builtins.elem stdenv.system strace.meta.platforms) [
```
Comment (hraban) [inline pkgs/development/compilers/sbcl/default.nix:93]: `stdenv.system` or `stdenv.hostPlatform.system`?

Comment (7c6f434c) [reply pkgs/development/compilers/sbcl/default.nix:93]: Hmmm, do we even test in cross-compilation anyway?

**Comment by 7c6f434c** [comment]
Still needs more disabled tests on Darwin…

**Comment by hraban** [comment]
:/ I'm happy to disable more and more, but these errors seem to occur randomly across different tests every other run. I don't think disabling a specific test is helping much in this case.

I have gone ahead and disabled testing for x86_64-darwin entirely. It is still an improvement over the status quo which is "no testing at all for any platform", not to mention that x86_64-darwin is a dying platform, so what does it really matter..? 🤷 

Let's see what this yields.

**Comment by 7c6f434c** [comment]
Obviously, I will _gladly_ defer to you for any Darwin-only decisions! (Given me not having any macOS devices around and all that)

**Comment by hraban** [comment]
Unless Apple invests in Intel hardware going forward, I don't see the point in spending much time on x86_64-darwin. The last Intel machine in their entire product line was the 2020 iMac.

I'll fix something if it's broken because I like my old laptop, but that's about it.

File: pkgs/development/compilers/sbcl/default.nix (hunk 2 above)
Comment (7c6f434c) [reply pkgs/development/compilers/sbcl/default.nix:93]: Correction: are we ever going to have working SBCL cross-compilation anyway, given that the upstream level of support is «well you need to run _some_ parts of the process on the host arch and not on the build arch anyway». (My bet is not until a large — and not too likely — change upstream)

**Comment by 7c6f434c** [comment]
@ofborg build sbcl.pkgs.stumpwm nyxt

**Comment by 7c6f434c** [comment]
I am not sure how such an ofborg failure is aarch64-darwin-only, but it looks plausibly legitimate…

**Comment by hraban** [comment]
I think aarch64-darwin ofborg is going through a bit of a rough patch. I opened another patch around the same time and its aarch64-darwin run has been queued for 1½ day now.

I didn't change anything on aarch64-darwin since it last passed and it was the most stable platform to begin with so I'm confident it's ok. Worst case we can disable it again if it goes wrong on hydra.

**Comment by 7c6f434c** [comment]
I think it fails even before the tests! But maybe there is an out-of-space issue that manifests weirdly.

**Pull request merged**